### PR TITLE
Make IHostEnvironment.ApplicationName non-nullable

### LINKED ISSUE #68512: [API Proposal]: Make IHostEnvironment.ApplicationName non-nullable
### Background and motivation

In 7.0-preview3 we made `IHostEnvironment.ApplicationName` and `IHostingEnvironment.ApplicationName` nullable because of the following:

https://github.com/dotnet/runtime/blob/05cb7f5a6eefbcd0fbab3bb5924c07b188d12501/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L202-L213

https://github.com/dotnet/runtime/pull/65403/files/dabdada197861a1fe343f0c82e6ebf3f87c663b5#r829310348

While this is technically correct, it's not user friendly. Historically, applications have expected this property to be non-nullable like the other properties on the interface like `EnvironmentName`.

It would be better if, similar to `EnvironmentName` and "Production", we just pick a default `ApplicationName` when `Assembly.GetEntryAssembly()` returns `null`.

### API Proposal

```diff
namespace Microsoft.Extensions.Hosting;

 public interface IHostEnvironment {
-    string? ApplicationName { get; set; }
+    string ApplicationName { get; set; }
 }
 public interface IHostingEnvironment {
-    string? ApplicationName { get; set; }
+    string ApplicationName { get; set; }
 }
```

### API Usage

Instead of doing complicated null-checks:

```csharp
if (env.ApplicationName is not null)
{
    var appAssembly = Assembly.Load(new AssemblyName(env.ApplicationName));
    if (appAssembly != null)
    {
        config.AddUserSecrets(appAssembly, optional: true);
    }
}
```

We should be able to assume `ApplicationName` is non-null:

```csharp
var appAssembly = Assembly.Load(new AssemblyName(env.ApplicationName));
if (appAssembly != null)
{
    config.AddUserSecrets(appAssembly, optional: true);
}
```

See https://github.com/dotnet/aspnetcore/pull/40754/files#diff-ba47e8a57bffee356b71c57b0ac4709dc0eb1329d547a36ee30fea262cf8358c

### Alternative Designs

Keeping `ApplicationName` nullable.

### Risks

We cannot come up with a good `ApplicationName` for when `Assembly.GetEntryAssembly()` returns `null`.

### PR DESCRIPTION
**Pull request by eerhardt**
It is not intended that normal execution has a null ApplicationName, so don't make callers need to check for null. When an ApplicationName is not available, it is set to string.Empty.

Fix #68512

### CODE REVIEW

**Review by tarekgh — APPROVED** [review]

**Comment by eerhardt** [comment]
@halter73 - any concern with using `string.Empty` instead of `<Unknown>` here?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/HostingEnvironment.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ public class HostingEnvironment : IHostingEnvironment, IHostEnvironment
     {
         public string EnvironmentName { get; set; } = null!;
 
-        public string? ApplicationName { get; set; }
+        public string ApplicationName { get; set; } = string.Empty;
```
Comment (halter73) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/HostingEnvironment.cs:18]: Nit: Should we also update `EnvironmentName` to initially be set to "Production" for consistency in how we implement the non-nullable default value?

Or update the assignment of this to be `ApplicationName = Assembly.GetEntryAssembly()?.GetName().Name ?? string.Empty` for consistency the other way?

**Review by halter73 — APPROVED** [review]
I think using `string.Empty` instead of something like `<Unknown>` is reasonable since hosting assumes that a non-empty app name is the name of the assembly containing the `UserSecretsIdAttribute`. It's easier to special case `string.Empty` than `<Unknown>`, and still communicates that a good default app name couldn't be determined.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/HostingEnvironment.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/HostingEnvironment.cs:18]: I don't really like having "policy" here... so instead to be consistent I changed all the non-nullable `string`s in this class to default to `string.Empty`.

**Pull request merged**